### PR TITLE
Enable multiple trace display in notebooks

### PR DESCRIPTION
**Pull request by daniellok-db**
### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?

This PR refactors all the display logic into a singleton display client. Since we need to call `display_traces()` from multiple parts of the codebase (e.g. from `search_traces()`), I felt it made more sense to remove the logic from the trace client.

The display client keeps track of the current IPython execution count in order to make sure the trace UI only shows up one time. The traces are deduped by `request_id` before being passed to the display function, so multiple copies of the same trace will not be displayed in the UI.

The way this works is similar to how we display a single trace—we just call ipython's native display with a custom mimebundle.

**NOTE 1**: this PR can only be merged after the notebook UI is updated to handle a list of traces. otherwise the UI will crash.

**NOTE 2**: the output of `get_traces()` will still be displayed in the cell if it was the last evaluated thing. I'm not sure if there's a good way to suppress this for now, because it's native IPython behavior to display the last evaluated object:

Screenshot 2024-04-05 at 5 42 24 PM

### How is this PR tested?

- [ ] Existing unit/integration tests
- [ ] New unit/integration tests
- [x] Manual tests

I will add some automated tests to this PR, but for any early reviewers:

https://github.com/mlflow/mlflow/assets/148037680/af99bd37-d0ed-4682-a2cc-335e5c4139e5

### Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [x] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/tracing/fluent.py [hunk 1]
```diff
@@ -232,12 +233,16 @@ def pagination_wrapper_func(number_to_get, next_page_token):
...
         )
 
-    return get_results_from_paginated_fn(
+    results = get_results_from_paginated_fn(
         pagination_wrapper_func,
         max_results_per_page=SEARCH_TRACES_DEFAULT_MAX_RESULTS,
         max_results=max_results,
     )
 
+    get_display_client().display_traces(results)
```
Comment (harupy) [inline mlflow/tracing/fluent.py:242]: what would happen if `results` contains 1000 items?

File: mlflow/tracing/display/display_client.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+def _serialize_trace_list(traces: List[Trace]):
+    return json.dumps(
+        # we can't just call trace.to_json() because this
+        # will cause the trace to be serialized twice (once
+        # by to_json and once by json.dumps)
+        [json.loads(trace.to_json()) for trace in traces]
+    )
+
+
+class IPythonTraceDisplayClient:
```
Comment (B-Step62) [inline mlflow/tracing/display/display_client.py:16]: I agree with extracting the display logic from the client, but the term `Client` seems to be a bit overloaded. We have three layers of client (1) MlflowClient (2) TraceClient and now (3) DisplayClient. Can we rename it to sth else like `DisplayHandler`?

File: mlflow/tracing/display/display_client.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+        self.display_handle = None
+        self.traces_to_display = {}
+
+    def get_mimebundle(self, traces: List[Trace]):
+        if len(traces) == 1:
+            return traces[0]._repr_mimebundle_()
+        else:
+            return {
+                "application/databricks.mlflow.trace": _serialize_trace_list(traces),
+                "text/plain": traces.__repr__(),
```
Comment (harupy) [inline mlflow/tracing/display/display_client.py:36]:
```suggestion
                "text/plain": repr(traces),
```

File: mlflow/tracing/display/display_client.py (hunk 2 above)
Comment (daniellok-db) [reply mlflow/tracing/display/display_client.py:16]: Yeah I get what you mean, we have a lot of clients haha. I'll make the update!

File: mlflow/tracing/display/display_client.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+                    self.get_mimebundle(deduped_trace_list),
+                    display_id=True,
+                    raw=True,
+                )
+        except Exception:
+            # swallow exceptions. this function is called as
+            # a side-effect in a few other functions (e.g. log_trace,
+            # get_traces, search_traces), and we don't want to block
+            # the core functionality if the display fails.
+            pass
```
Comment (B-Step62) [inline mlflow/tracing/display/display_client.py:78]: nit: Can we show warning or info here? I remember it was a bit challenging to debug the "trace not displayed" issue during bug-bash:)

Comment (BenWilson2) [reply mlflow/tracing/display/display_client.py:78]: Let's use _logger.debug() to render the exception

File: mlflow/tracing/clients/local.py [hunk 5]
```diff
@@ -93,6 +65,7 @@ def get_trace(self, request_id: str) -> Optional[Trace]:
         with self._lock:
             for trace in self.queue:
                 if trace.trace_info.request_id == request_id:
+                    get_display_handler().display_traces([trace])
```
Comment (B-Step62) [inline mlflow/tracing/clients/local.py:68]: Can we move this call to upstream i.e. the user-facing APIs (`mlflow.MlflowClient().get_trace()`, `mlflow.get_traces`), and make this in-memory client focus purely on *getting* traces without side-effect? These getter APIs are also used internally to access to in-memory trace like [here](https://github.com/mlflow/mlflow/blob/tracing/mlflow/tracking/client.py#L624).

Comment (daniellok-db) [reply mlflow/tracing/clients/local.py:68]: makes sense, i'll update!

**Review by B-Step62 — APPROVED** [review]
LGTM! Thanks for the quick update!

**Pull request merged**